File: src/main/java/me/santio/utils/inventories/CustomInventory.java
package me.santio.utils.inventories;

import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import me.santio.utils.ChatUtils;
import me.santio.utils.SantioUtils;
import me.santio.utils.inventories.buttons.Button;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.metadata.FixedMetadataValue;
import org.bukkit.util.Consumer;

import java.util.HashMap;
import java.util.List;
import java.util.Optional;

@SuppressWarnings({"unused", "StringConcatenationInLoop", "UnusedReturnValue"})
@Accessors(chain = true)
public final class CustomInventory {
    @Getter private final SantioUtils utils;
    
    @Setter @Getter private boolean deleteOnClose = true;
    @Getter private final int size;
    @Getter private final String name;
    @Getter private final Inventory inventory;
    @Getter private String ID;
    
    @Getter private final HashMap<String, Consumer<InventoryClickEvent>> events = new HashMap<>();
    
    public CustomInventory(int size, String name, SantioUtils utils) {
        this.utils = utils;
        
        this.size = (size % 9 == 0) ? size : 9 * size;
        this.name = name;
        this.ID = ChatUtils.toID(name);
        while (utils.getInventories().containsKey(getID())) this.ID += "_1";
        
        inventory = Bukkit.createInventory(null, this.size, ChatUtils.tacc(name));
        utils.getInventories().put(getID(), this);
    }
    
    public CustomInventory open(Player... players) {
        for (Player player : players) open(player);
        return this;
    }
    
    public CustomInventory open(Player player) {
        player.openInventory(getInventory());
        player.setMetadata("sutils_inventory_open", new FixedMetadataValue(utils.getPlugin(), getID()));
        return this;
    }
    
    public boolean isOpen(Player player) {
        return player.hasMetadata("sutils_inventory_open") && (player.getMetadata("sutils_inventory_open").get(0).asString().equals(getID()));
    }
    
    public void delete() {
        for (Player player : Bukkit.getOnlinePlayers()) {
            if (isOpen(player)) player.closeInventory();
        }
        events.clear();
        utils.getInventories().remove(getID());
    }
    
    public CustomInventory setItem(int slot, Button item) {
        return setItem(slot, item, (e) -> {
            e.setCancelled(true);
            item.onClick(e);
            updateLore(slot, item.getCompleteLore());
        });
    }
    
    public CustomInventory updateLore(int slot, List<String> lore) {
        ItemStack item = inventory.getItem(slot);
        if (item == null) return this;
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return this;
        
        meta.setLore(lore);
        item.setItemMeta(meta);
        inventory.setItem(slot, item);
        return this;
    }
    
    public CustomInventory setItem(int slot, Button item, Consumer<InventoryClickEvent> event) {
        ItemStack itemStack = item.getItem();
        ItemMeta meta = itemStack.getItemMeta();
        
        if (meta != null) {
            String id = "event-" + getEvents().size();
            utils.NBTUtils.set(meta, "sutils_id", id);
            meta.setLore(item.getCompleteLore());
            itemStack.setItemMeta(meta);
            getEvents().put(id, event);
        }
    
        getInventory().setItem(slot, itemStack);
        return this;
    }
    
    public CustomInventory setItem(int slot, ItemStack item, Consumer<InventoryClickEvent> event) {
        ItemMeta meta = item.getItemMeta();
        if (meta != null) {
            String id = "event-" + getEvents().size();
            utils.NBTUtils.set(meta, "sutils_id", id);
            item.setItemMeta(meta);
            getEvents().put(id, event);
        }
        
        getInventory().setItem(slot, item);
        return this;
    }
    
    public CustomInventory setItem(int slot, ItemStack item) {
        return setItem(slot, item, (e) -> e.setCancelled(true));
    }
    
    public CustomInventory fill(ItemStack item) {
        for (int i = 0; i<getSize(); i++) setItem(i, item);
        return this;
    }
    
    public CustomInventory border(ItemStack item) {
        for (int i = 0; i<9; i++) setItem(i, item);
        for (int line = 0; line < size / 9D; line++) {
            setItem(line * 9, item);
            setItem((int) ((line * 9D) + 8), item);
        }
        for (int i = (size - 10); i<(size - 1); i++) setItem(i, item);
        return this;
    }
    
    public static Optional<CustomInventory> getInventory(SantioUtils utils, Player player) {
        return utils.getInventories().values()
                .stream()
                .filter((CustomInventory i) -> i.isOpen(player))
                .findFirst();
    }
    
}

File: src/main/java/me/santio/utils/inventory/CustomInventory.kt
package me.santio.utils.inventory

import me.santio.utils.SantioUtils
import me.santio.utils.item.CustomItem
import me.santio.utils.text.colored
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import java.util.*
import java.util.function.Consumer
import java.util.function.Function

@Suppress("MemberVisibilityCanBePrivate", "unused")
open class CustomInventory @JvmOverloads constructor(open val size: Int, var name: String = "Inventory") {
    constructor(inventory: CustomInventory): this(inventory.size, inventory.name)

    companion object {
        @JvmStatic
        fun getOpenInventory(player: Player): CustomInventory? {
            for (inventory in SantioUtils.inventories) {
                if (inventory.isOpen(player)) {
                    return inventory
                }
            }
            return null
        }

        @JvmStatic
        fun get(inventory: Inventory): CustomInventory? {
            for (customInventory in SantioUtils.inventories) {
                if (customInventory.inventory == inventory) {
                    return customInventory
                }
            }
            return null
        }
    }

    private var deleteOnClose: Boolean = true
    protected var inventory: Inventory = Bukkit.createInventory(null, size(), name.colored())
    protected var onClick: MutableMap<Int, Consumer<CustomInventoryClickEvent>> = mutableMapOf()
    internal var previous: CustomInventory? = null
    internal var next: CustomInventory? = null
    val id: String = UUID.randomUUID().toString()

    fun getBukkitInventory() = inventory
    fun rows(): Int = if (size % 9 == 0) size / 9 else size
    fun size(): Int = rows() * 9
    fun items(): List<ItemStack> = inventory.contents.toList()
    fun opened(): List<Player> = inventory.viewers.map { it as Player }
    fun isEmpty(ignore: Player? = null): Boolean = opened().none { it != ignore }

    fun isOpen(player: Player): Boolean = opened().contains(player)

    fun deleteOnClose(deleteOnClose: Boolean): CustomInventory {
        this.deleteOnClose = deleteOnClose
        return this
    }

    fun deleteOnClose() = deleteOnClose

    fun onClick(event: InventoryClickEvent) {
        onClick[event.slot]?.accept(CustomInventoryClickEvent(event, this)) ?: run { event.isCancelled = true }
    }

    open fun rename(name: String): CustomInventory {
        this.name = name

        val export = export()
        inventory = Bukkit.createInventory(null, size(), name.colored())

        import(export)
        open(*opened().toTypedArray())

        return this
    }

    @JvmOverloads
    fun set(
        slot: Int,
        item: ItemStack,
        onClick: Consumer<CustomInventoryClickEvent>? = null,
        keepEvent: Boolean = false
    ): CustomInventory {
        inventory.setItem(slot, item)
        if (onClick != null) this.onClick[slot] = onClick
        else if (!keepEvent) this.onClick.remove(slot)
        return this
    }

    @JvmOverloads
    fun set(
        slot: Slots,
        item: ItemStack,
        onClick: Consumer<CustomInventoryClickEvent>? = null
    ): CustomInventory {
        slot.apply(this).forEach { set(it, item, onClick) }
        return this
    }

    @JvmOverloads
    fun add(
        item: ItemStack,
        onClick: Consumer<CustomInventoryClickEvent>? = null
    ): CustomInventory {
        val slot = inventory.firstEmpty()
        if (slot == -1) return this

        set(slot, item, onClick)
        return this
    }

    fun clear(vararg slots: Int): CustomInventory {
        slots.forEach {
            inventory.setItem(it, null)
            onClick.remove(it)
        }

        return this
    }

    @JvmOverloads
    fun clear(slots: Slots = Slots.ALL): CustomInventory {
        slots.apply(this).forEach { clear(it) }
        return this
    }

    @JvmOverloads
    fun border(
        item: ItemStack,
        onClick: Consumer<CustomInventoryClickEvent> = Consumer { it.isCancelled = true }
    ): CustomInventory {
        val slots = Slots.column(1)
            .add(Slots.column(9))
            .add(Slots.row(1))
            .add(Slots.row(rows()))

        set(slots, item, onClick)
        return this
    }

    fun export(): List<SlotData> {
        val slots = mutableListOf<SlotData>()
        for (i in 0 until size()) {
            val item = inventory.getItem(i)
            if (item != null) {
                slots.add(SlotData(i, item, onClick[i]))
            }
        }

        return slots
    }

    fun import(slots: List<SlotData>): CustomInventory {
        slots.forEach { set(it.slot, it.item, it.clickEvent) }
        return this
    }

    fun open(vararg players: Player): CustomInventory {
        players.forEach {
            it.openInventory(inventory)
        }

        SantioUtils.inventories.add(this)
        return this
    }

    fun delete() {
        if (next != null) return
        SantioUtils.inventories.remove(this)
        opened().forEach { it.closeInventory() }
    }

    @JvmOverloads
    fun <T: Enum<T>> paginate(
        slots: Slots,
        clazz: Class<T>,
        handler: Function<T, CustomItem>,
        event: Function<T, Consumer<CustomInventoryClickEvent>> = Function { Consumer { it.isCancelled = true } },
        filter: ((T) -> Boolean)? = null
    ): PaginatedInventory {
        val paginatedInventory = PaginatedInventory(this)
        paginatedInventory.useSlots(slots)
        paginatedInventory.addItems(*clazz.enumConstants.filter(filter ?: { true }).map { handler.apply(it) }.toTypedArray())
        clazz.enumConstants.filter(filter ?: { true }).map { event.apply(it) }.forEachIndexed { index, consumer ->
            slots.get(index)?.let {
                paginatedInventory.onClick[it] = consumer
                paginatedInventory.setEvent(index / slots.size() + 1, it, consumer)
            }
        }
        return paginatedInventory
    }

    fun paginate(
        slots: Slots,
        items: List<CustomItem>,
    ): PaginatedInventory {
        val paginatedInventory = PaginatedInventory(this)
        paginatedInventory.useSlots(slots)
        paginatedInventory.addItems(*items.toTypedArray())

        items.forEachIndexed { index, item ->
            val page = index / slots.size()
            paginatedInventory.setEvent(page, index % slots.size()) {
                item.onClick(it)
            }
        }

        return paginatedInventory
    }

}